cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DefaultDatasetRuntimeContext.java

```diff
@@ -119,44 +125,45 @@ protected CallStack childValue(CallStack parentValue) {
   }
 
   private DefaultDatasetRuntimeContext(AuthorizationEnforcer enforcer, DatasetAccessRecorder accessRecorder,
-                                       Principal principal, DatasetId datasetId) {
+                                       Principal principal, DatasetId datasetId,
+                                       @Nullable Class<? extends Annotation> constructorDefaultAnnotation) {
     this.enforcer = enforcer;
     this.accessRecorder = accessRecorder;
     this.principal = principal;
     this.datasetId = datasetId;
+    this.constructorDefaultAnnotation = constructorDefaultAnnotation;
     this.lineageRecorded = new boolean[AccessType.values().length];
     this.auditRecorded = new boolean[AccessType.values().length];
   }
 
   @Override
-  public void onMethodEntry(@Nullable Class<? extends Annotation> annotation) {
+  public void onMethodEntry(boolean constructor, @Nullable Class<? extends Annotation> annotation) {
     CallStack callStack = this.callStack.get();
-    AccessInfo accessInfo;
-    AccessType lineageType;
-    AccessType auditType;
+    AccessInfo accessInfo = UNKNOWN_ACCESS_INFO;
+
+    if (annotation == null && constructor) {
+      annotation = constructorDefaultAnnotation;
+    }
 
     if (annotation != null) {
       accessInfo = ANNOTATION_TO_ACCESS_INFO.get(annotation);
       if (accessInfo == null) {
         // shouldn't happen
         throw new DataSetException("Unsupported annotation " + annotation + " on dataset " + datasetId);
       }
+    }
 
-      // Perform authorization if the method is annotated.
-      try {
-        enforcer.enforce(datasetId, principal, accessInfo.getActions());
-      } catch (Exception e) {
-        throw new DataSetException("The principal " + principal + " is not authorized to access " + datasetId +
-                                     " for operation types " + accessInfo.getActions(), e);
-      }
-
-      lineageType = callStack.enter(accessInfo.getAccessType());
-      auditType = accessInfo.getAccessType();
-    } else {
-      lineageType = callStack.enter(AccessType.UNKNOWN);
-      auditType = AccessType.UNKNOWN;
+    // Performs authorization.
+    // If the method is not annotated, the action set is empty, which means the user need to have some privileges,
+    // but we won't allow no privilege at all
+    try {
+      enforcer.enforce(datasetId, principal, accessInfo.getActions());
+    } catch (Exception e) {
+      throw new DataSetException("The principal " + principal + " is not authorized to access " + datasetId +
```

Comment:
The exception will shown the action set as an empty set `[]`. But I think it's ok, unless want to specially handle the empty set case. We use empty set to check for if the user has any access right.